### PR TITLE
Inner join distinction

### LINKED ISSUE #1238: [API Improvement] inner/outer join distinction
Hi,

this issue is a proposal of small API improvement.

Typically in databases/other processing systems use "join" as an alias to "inner join". That's the reason I was suprised that `hashJoin2` produced tuple with null. In the docs we have info that join is asymmetric, but it's still not intuitive ;)

My proposal is to:
- deprecate hashJoin/hashJoin2
- add function leftJoin / rightJoin
- add function join or innerJoin (which can be implemented as leftJoin + filter not null)
- make hashjoinbuilder more fluent api, so that we won't have to split  code into many variables, but write it in more functional way.

Is it possible? What do you think? In my opinion it will make API cleaner and easier

### PR DESCRIPTION
**Pull request by TomaszGaweda**
As discussed on Gitter and mentioned in #1238, new users often don't know that hashJoin is not inner join (like their intuition says). Adding innerHashJoin with pre-filtering will reduce filtering in user apps and make API more clear for new users.

Checklist
- [x] Tags Set
- [x] Milestone Set
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc
- [x] No breaking changes

Fixes #1238

### CODE REVIEW

**Comment by TomaszGaweda** [comment]
I'm thinking if left and right join methods also should be added. Currently innerJoin method will reduce a lot number of custom filtering methods (and will make Java 14's records even more enjoyable :) ), not sure if left join is also so commonly used and if we should add another methods to public API

**Comment by TomaszGaweda** [comment]
@cangencer @viliam-durina Can I ask you for a review of this small patch? :)

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 1]
```diff
@@ -750,6 +750,48 @@
...
+     * BatchStage<Map.Entry<Long, Country>> idAndCountry;
+     *
+     * users.innerHashJoin(
+     *     idAndCountry,
+     *     JoinClause.joinMapEntries(User::getCountryId),
+     *     (user, country) -> user.setCountry(country)
+     * )
+     * }</pre>
+     *
+     * <p>
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java]: Please follow the formatting guidelines you can infer from the rest of the docs. We use text line length of 72 (excluding any indentation and `*`),  we don't use `</p>` and we don't indent the text.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 2]
```diff
@@ -750,6 +750,48 @@
...
+     *     This metod is similar to {@link #hashJoin(BatchStage, JoinClause, BiFunctionEx)} method,
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java]: Typo: `metod`

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 3]
```diff
@@ -750,6 +750,48 @@
             @Nonnull BiFunctionEx<T, T1, R> mapToOutputFn
     );
 
+    /**
+     * Attaches to both this and the supplied stage a inner-hash-joining stage and
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java]: Typo: `a inner`, should be `an inner`.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java (hunk 1 above)
Comment (mtopolnik) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java]: You should also update the Javadoc that this doc points to. The `com.hazelcast.jet.pipeline` Javadoc now explicitly states that "hash join is a left-outer join".

**Comment by TomaszGaweda** [comment]
Thanks @mtopolnik, all comments are addresses now

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java [hunk 4]
```diff
@@ -72,12 +72,25 @@
  * hashtables (hence the name). It consumes the enriching streams in full
  * before ingesting any data from the primary stream.
  * <p>
- * The output is just like an SQL left-outer join: for each primary item
- * there are N output items, one for each matching item in the enriching
- * set. If an enriching set doesn't have a matching item, the output will
- * have a {@code null} instead of the enriching item. The join also allows
+ * The output of {@code hashJoin} is just like an SQL left-outer join:
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java]: Please reflow these paragraphs.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java [hunk 5]
```diff
@@ -72,12 +72,25 @@
...
+ * for each primary item there are N output items,
+ * one for each matching item in the enriching set.
+ * If an enriching set doesn't have a matching item, the output will
+ * have a {@code null} instead of the enriching item.
+ * <p>
+ * If SQL inner-join is needed,
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java]: Please try to use the active voice as much as possible. This can be "If you need the inner join, ". Also note "inner join" (without a dash). We actually need to fix "hash-join" into "hash join" as well, but this is not the topic of this PR.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java [hunk 6]
```diff
@@ -72,12 +72,25 @@
...
+ * then specialised {@code innerHashJoin} can be used, in which
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java]: _the_ specialized innerHashJoin

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java [hunk 7]
```diff
@@ -72,12 +72,25 @@
...
+ * for each primary item with at least one match, there are N output items,
+ * one for each matching item in the enriching set.
+ * If an enriching set doesn't have a matching item,
+ * there will be no records with given primary item.
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java]: with _the_ given primary item

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java [hunk 8]
```diff
@@ -72,12 +72,25 @@
...
+ * In this case output function's arguments are always non-null.
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java]: _the_ output function's

File: hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/BatchStageTest.java [hunk 9]
```diff
@@ -1035,6 +1035,35 @@ public void hashJoin_when_outputFnReturnsNull_then_filteredOut() {
         assertEquals(emptyList(), new ArrayList<>(sinkList));
     }
 
+    @Test
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/BatchStageTest.java]: This claims to test that `innerHashJoin` filters out nulls, but it actually uses `hashJoin` and asserts that it does not filter out nulls.
About method naming: we use `_` only for special purposes like the structure `when_predondition_then_postcondition`. So this method should be `when_hashJoin_then_dontFilterOutNulls` or similar.

File: hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/StreamStageTest.java [hunk 10]
```diff
@@ -1253,6 +1253,39 @@ public void hashJoin() {
         );
     }
 
+    @Test
+    @SuppressWarnings("unchecked")
+    public void innerHashJoin() {
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/StreamStageTest.java]: This seems to test the behavior common to inner and outer join. The input has no nulls to begin with.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/BatchStageImpl.java [hunk 11]
```diff
@@ -165,6 +165,21 @@ public BatchStageImpl(@Nonnull Transform transform, FunctionAdapter ignored, @No
...
+    ) {
+        BiFunctionEx<T, T1, R> finalOutputFn = (leftSide, rightSide) -> {
+            if (leftSide == null || rightSide == null) {
+                return null;
+            }
+            return mapToOutputFn.apply(leftSide, rightSide);
+        };
+        return attachHashJoin(stage1, joinClause1, finalOutputFn);
+    }
+
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/BatchStageImpl.java]: We also have `hashJoin2()` and `hashJoinBuilder()`, they also need the same API extension. You could add a method to the hash join builder: `buildInner`.

**Comment by cangencer** [comment]
@TomaszGaweda once you have done the javadoc updates we can merge this.

**Comment by TomaszGaweda** [comment]
Hi, first of all - thanks for all the comments @mtopolnik! I agree with them and will fix it. Secondly, @cangencer I will try to make changes in next 2 days, I didn't manage to get some time to implement them. Thanks for your patience

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java (hunk 4 above)
Comment (TomaszGaweda) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java:75]: > 
> 
> Please reflow these paragraphs.

By reflow you mean only formatting or also how sentences are written?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/BatchStageImpl.java (hunk 11 above)
Comment (TomaszGaweda) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/BatchStageImpl.java:182]: Maybe not builderInner, but rather addInner? So you can have left outer and inner join in one builder

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java (hunk 4 above)
Comment (mtopolnik) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/package-info.java]: Just the formatting aspect :)

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/BatchStageImpl.java (hunk 11 above)
Comment (mtopolnik) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/pipeline/BatchStageImpl.java:182]: Yes, sounds better than my idea.

**Comment by cangencer** [comment]
@mtopolnik could you re-review? @TomaszGaweda thanks a lot for the contribution!

**Comment by cangencer** [comment]
Also this page will need updating: https://jet-start.sh/docs/api/stateless-transforms

**Comment by cangencer** [comment]
@mtopolnik could you summarize if anything is still missing on this PR?

**Comment by mtopolnik** [comment]
There are two unresolved conversations, those are the missing bits. One is just about proper Javadoc formatting, the other is about the missing inner join support for `hashJoin2` and `hashJoinBuilder()`.

**Comment by TomaszGaweda** [comment]
JavaDocs should be ok now, I hope to make changes to innerHashJoin2 and HashJoinBuilder this week - sorry, was quite busy and didn't have time to look at it

**Comment by TomaszGaweda** [comment]
@mtopolnik all changes should be implemented now, can you please recheck? :)

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 12]
```diff
@@ -750,6 +750,49 @@
             @Nonnull BiFunctionEx<T, T1, R> mapToOutputFn
     );
 
+    /**
+     * Attaches to both this and the supplied stage an inner hash-joining stage
+     * and  returns it. This stage plays the role of the <em>primary stage</em>
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java:755]: There's a double space here.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 13]
```diff
@@ -795,6 +838,58 @@
             @Nonnull TriFunction<T, T1, T2, R> mapToOutputFn
     );
 
+    /**
+     * Attaches to this and the two supplied stages a inner hash-joining stage and
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java:842]: Line too long.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 14]
```diff
@@ -795,6 +838,58 @@
...
+     *
+     * users.innerHashJoin2(
+     *     idAndCountry, JoinClause.joinMapEntries(User::getCountryId),
+     *     idAndCompany, JoinClause.joinMapEntries(User::getCompanyId),
+     *     (user, country, company) -> user.setCountry(country).setCompany(company)
+     * )
+     * }</pre>
+     *
+     * <p>
+     * This method is similar to {@link #hashJoin2(BatchStage, JoinClause, BatchStage, JoinClause, TriFunction)}
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java:864]: Line too long.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 15]
```diff
@@ -750,6 +750,49 @@
...
+     *
+     * users.innerHashJoin(
+     *     idAndCountry,
+     *     JoinClause.joinMapEntries(User::getCountryId),
+     *     (user, country) -> user.setCountry(country)
+     * )
+     * }</pre>
+     *
+     * <p>
+     * This method is similar to {@link #hashJoin(BatchStage, JoinClause, BiFunctionEx)}
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java:774]: Line too long. Note that you can just refer to a method name without parameters.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralHashJoinBuilder.java [hunk 16]
```diff
@@ -82,7 +82,28 @@
...
         Tag<T1> tag = tag(clauses.size());
-        clauses.put(tag, new TransformAndClause<>(stage, joinClause));
+        clauses.put(tag, new TransformAndClause<>(stage, joinClause, false));
+        return tag;
+    }
+
+    /**
+     * Adds another contributing pipeline stage to the hash-join operation.
+     *
+     * If no matching items for returned {@linkplain Tag tag} is found, no records
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralHashJoinBuilder.java:92]: Line too long.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 17]
```diff
@@ -750,6 +750,49 @@
...
+     * method, but it guarantees that both input items will be not-null. Nulls will
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java:775]: Line too long.

File: hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/BatchStageTest.java [hunk 18]
```diff
@@ -1035,6 +1035,75 @@ public void hashJoin_when_outputFnReturnsNull_then_filteredOut() {
...
+                        .flatMap(i -> traverseItems(entry(i, prefixA + i), entry(i, prefixB + i)));
+        BatchStage<Entry<Integer, String>> enrichingStage2 =
+                batchStageFromList(input)
+                        .filter(e -> e <= itemCountLocal / 4)
+                        .flatMap(i -> traverseItems(entry(i, prefixC + i), entry(i, prefixD + i)));
+
+        // When
+        HashJoinBuilder<Integer> b = batchStageFromList(input).hashJoinBuilder();
+        Tag<String> tagA = b.addInner(enrichingStage1, joinMapEntries(wholeItem()));
+        Tag<String> tagB = b.addInner(enrichingStage2, joinMapEntries(wholeItem()));
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/BatchStageTest.java:1084]: Sure. BTW. One question - typically when we do
x 
left join y
join z

then join to y is also treated as inner join. Currently only no value on z will cause no tuple returned, the lack of y will be ignored. Is it ok or should we stay as compatible with SQL as possible?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralHashJoinBuilder.java [hunk 19]
```diff
@@ -98,17 +119,23 @@
                         orderedClauses.stream().map(e -> e.getValue().transform())
                 ).collect(toList());
         // A probable javac bug forced us to extract this variable
+        // and not using method reference
         Stream<JoinClause<?, T0, ?, ?>> joinClauses = orderedClauses
                 .stream()
                 .map(e -> e.getValue().clause())
-                .map(fnAdapter::adaptJoinClause);
+                .map(joinClause -> fnAdapter.adaptJoinClause(joinClause));
+        Stream<Boolean> whereNullNotAllowed = orderedClauses
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralHashJoinBuilder.java:127]: I tried it, you can safely inline this into the usage below. The reason that `joinClauses` stream had to be extracted does not appear in this case.

File: hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/BatchStageTest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (mtopolnik) [reply hazelcast-jet-core/src/test/java/com/hazelcast/jet/pipeline/BatchStageTest.java:1084]: Treating each side separately seems more flexible and useful to me. You may legitimately need such a combination. SQL strictly follows relational algebra, and there we'd consider the whole expression as a composition of two joins, so if the first one is an inner join, the tuple is eliminated even before reaching the left join. Jet's transform are not an implementation of relational algebra so I'm fine with going the more useful route.

 @viliam-durina what do you think?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralHashJoinBuilder.java (hunk 19 above)
Comment (TomaszGaweda) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralHashJoinBuilder.java]: Done

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java [hunk 20]
```diff
@@ -795,6 +838,58 @@
...
+     * users.innerHashJoin2(
+     *     idAndCountry, JoinClause.joinMapEntries(User::getCountryId),
+     *     idAndCompany, JoinClause.joinMapEntries(User::getCompanyId),
+     *     (user, country, company) -> user.setCountry(country).setCompany(company)
+     * )
+     * }</pre>
+     *
+     * <p>
+     * This method is similar to {@link #hashJoin2(BatchStage, JoinClause, BatchStage, JoinClause, TriFunction)}
+     * method, but it guarantees that both input items will be not-null. Nulls will
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/GeneralStage.java]: All this comments are addressed. Limit of 72 chars in line will be my next nightmare I think :D

**Comment by mtopolnik** [comment]
@TomaszGaweda Sorry for the inconvenience, but after introducing the automated check for contributors' license signing, it turned out that we couldn't convince the system some users already signed the agreement.

Can you please [http://ec2-54-211-157-189.compute-1.amazonaws.com:5000/hazelcast/hazelcast-jet?pullRequest=2089](use this link} to do it again? It is just point-and-click.

**Comment by TomaszGaweda** [comment]
No problem, done.

**Comment by Holmistr** [comment]
Hey @TomaszGaweda , I see that you actually signed the CLA, but the bot still thinks you didn't (https://github.com/hazelcast/hazelcast-jet/pull/2089#issuecomment-609433589). It says:

> Tomasz Gaweda seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please add the email address used for this commit to your account.

Would you mind checking this? I would like to learn where the problem is, so far it worked seamlessly.

**Comment by Holmistr** [comment]
@TomaszGaweda Possibly the email address that your local Git is configured with doesn't match your GH account one?

**Comment by TomaszGaweda** [comment]
@Holmistr It turned out, that yes, I had different email provided. I've changed my GitHub email, but cannot edit it on the CLA form

**Comment by Holmistr** [comment]
The comment above was actually me, wrong account :-D
